braindecode/models/usleep.py

```diff
@@ -0,0 +1,272 @@
+# Authors: Theo Gnassounou <theo.gnassounou@inria.fr>
+#          Omar Chehab <l-emir-omar.chehab@inria.fr>
+#
+# License: BSD (3-clause)
+
+import numpy as np
+import torch
+from torch import nn
+
+
+def _crop_tensors_to_match(x1, x2, axis=-1):
+    """Crops two tensors to their lowest-common-dimension along an axis."""
+    dim_cropped = min(x1.shape[axis], x2.shape[axis])
+
+    x1_cropped = torch.index_select(
+        x1, dim=axis,
+        index=torch.arange(dim_cropped).to(device=x1.device)
+    )
+    x2_cropped = torch.index_select(
+        x2, dim=axis,
+        index=torch.arange(dim_cropped).to(device=x1.device)
+    )
+    return x1_cropped, x2_cropped
+
+
+class _EncoderBlock(nn.Module):
+    """Encoding block for a timeseries x of shape (B, C, T)."""
+    def __init__(self,
+                 in_channels=2,
+                 out_channels=2,
+                 kernel_size=9,
+                 downsample=2):
+        super().__init__()
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.kernel_size = kernel_size
+        self.downsample = downsample
+        padding = (kernel_size - 1) // 2   # chosen to preserve dimension
+
+        self.block_prepool = nn.Sequential(
+                nn.Conv1d(in_channels=in_channels,
+                          out_channels=out_channels,
+                          kernel_size=kernel_size,
+                          padding=padding),
+                nn.ELU(),
+                nn.BatchNorm1d(num_features=out_channels),
+            )
+
+    def forward(self, x):
+        x = self.block_prepool(x)
+        residual = x
+        if x.shape[-1] % 2:
+            x = nn.ConstantPad1d(padding=1, value=0)(x)
+        x = nn.MaxPool1d(kernel_size=self.downsample, stride=self.downsample)(x)
```

Comment:
Why not just instantiate these during object construction like
```python
self.pad = nn.ConstantPad1d(padding=1, value=0)
self.pool = nn.MaxPool1d(kernel_size=self.downsample, stride=self.downsample)
```

Seems strange to me to create object to then just use it, then one could also call torch functional API?